### ISSUE TITLE
Allow attributes from YAML (or other structured configuration language) files

### ISSUE THREAD

**Issue by Crown0815** [asciidoctor/asciidoctor#4497]
The ability to define domain specific values in a generally available configuration language allows those values to be referenced in processing code and the documentation alike. 

We face many areas where domain experts are defining values that must be accessible through an application we are writing, but also in end-user documentation. For this purpose, it would be great to be able to define those values in something like YAML, or TOML, or JSON and refer to them from the `asciidoc` documentation and our code directly. 

Currently we would have to define one format and convert to another format within an automated build pipeline. Relying on a pipeline makes it significantly harder to check the correctness of the configuration file while working on the documentation and/or the code.

**Comment by mojavelinux** [comment]
This is a duplicate of #566.

To be quite honest, we are probably not going to do this in Asciidoctor core. Firstly, it can be achieved using a preprocessor extension (the preprocessor extension can contribute attributes to the API before the document is read and parsed). Secondly, it is something that I think is more fit for a static site generator, which is why Antora offers this feature. The other reason it's more suitable for a static site generator is that the tool needs more context than a single document, which is precisely what something like Antora gives you.

I'm not necessarily opposed to implementing #566, but there are still details to work out for it can be implemented. Those we may have a clearer picture now than we did before.

**Comment by Crown0815** [comment]
@mojavelinux I am not sure this is a duplicate of #566. 

While the wish to add compatibility for _traditional config file **formats**_ is similar, what I am proposing is to use those files for `include`_ish_ replacements (similar to how an `attributes.adoc` file would be used). 

The YAML/TOML/JSON file is not supposed to serve any `asciidoctor` configuration file, but purely as a store of attributes (in my case those serve as variables/constants for domain values).

The advantage of using one of these _config file formats_ would be that it simplifies sharing that information with other tools/languages, i.e., avoids duplication of information.

Regarding the `preprocessor extension`, where can I find documentation/examples on how to write one?

**Comment by mojavelinux** [comment]
Of course, it depends on how you look at it, but for what's being considered for Asciidoctor core, #566 covers it from that perspective.

First, I want to point out that the include directive adds lines to the reader to be parsed. That is how the language is defined and that is not going to change. However, Asciidoctor offers an extension point that allows you to intercept the include directive to add or replace functionality. There are also other extension points, such as a preprocessor or block macro processor, that would afford you other customizations.

Extensions are documented in the Asciidoctor documentation at https://docs.asciidoctor.org/asciidoctor/latest/extensions/. I've added an example to the extension lab that shows how to load a configuration file that contributes externalized attributes to the document header. See https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/main/lib/external-header-attributes-preprocessor.rb and https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/main/lib/external-header-attributes-preprocessor/extension.rb. I believe this will serve the needs you describe.

I believe if we implemented #566, this same behavior can be achieved without needing an extension, though it will be bound to the `asciidoctor` CLI.

Anther approach is to override the include directive and have it convert the included file into attribute entries which will be read by the processor (specifically the parser). Here's an example of how that can be done:

```rb
require 'yaml'

# intercepts include::shared-attributes.yml[] and adds attribute entries to the document
class AttributesIncludeProcessor < Asciidoctor::Extensions::IncludeProcessor
  def process doc, reader, target, attrs
    inc_path = doc.normalize_system_path target, reader.dir, nil, target_name: 'include file'
    if ::Hash === (attributes = YAML.load_file inc_path)
      lines = attributes.each_with_object([]) do |(name, val), lines|
        if val
          entry = %(:#{name}:)
          val = val.to_s
          entry += %( #{val}) if val.length
        else
          entry = %(:!#{name}:)
        end
        lines << entry
      end
      reader.push_include (lines.join ?\n), target, target, 1, attrs
    end
    reader
  end

  def handles? target
    target.end_with? '-attributes.yml'
  end
end

Asciidoctor::Extensions.register do
  include_processor AttributesIncludeProcessor
end
```

If you have further questions about this functionality or the extensions I shared, please raise your question in the project chat at https://chat.asciidoctor.org.

**Comment by mojavelinux** [comment]
I forgot to mention that a similar preprocessor extension can be found here:

* https://github.com/asciidoctor/asciidoctor-extensions-lab/blob/main/lib/git-metadata-preprocessor/extension.rb

In this case, it sets header attributes read from git metadata.
